kore/src/Kore/Exec.hs

```diff
@@ -282,20 +284,31 @@ search verifiedModule strategy termLike searchPattern searchConfig =
 
 -- | Proving a spec given as a module containing rules to be proven
 prove
-    ::  ( Log.WithLog Log.LogMessage smt
+    ::  forall smt claim
+      . ( Log.WithLog Log.LogMessage smt
         , MonadCatch smt
         , MonadProfiler smt
         , MonadUnliftIO smt
         , MonadSMT smt
+        , Claim claim
+        , Eq claim
+        , Show claim
+        , Show (Goal.Rule claim)
+        , Coercible claim (RulePattern Variable)
+        , TopBottom claim
+        , Goal.ProofState claim (Pattern Variable) ~ CommonProofState
         )
     => Limit Natural
     -> VerifiedModule StepperAttributes Attribute.Axiom
     -- ^ The main module
     -> VerifiedModule StepperAttributes Attribute.Axiom
     -- ^ The spec module
+    -> (VerifiedModule StepperAttributes Attribute.Axiom
+        -> [(Attribute.Axiom, claim)])
+    -- ^ claims extractor
     -> smt (Either (TermLike Variable) ())
-prove limit definitionModule specModule =
-    evalProver definitionModule specModule
+prove limit definitionModule specModule claimExtractor =
```

Comment:
Could we push the `claimExtractor` out into the `Claim` class?

Reply:
Thanks! I did that (kind of). However, I now need a `Proxy` for the `prove` and `proveWithRepl` functions as they have no `claim`-related arguments anymore. Vlad suggested we `AllowAmbiguousTypes` for the `Exec` module and use `TypeApplications`  instead of `Proxy` at call-time.  What do you think?

Reply:
> Vlad suggested we AllowAmbiguousTypes for the Exec module and use TypeApplications instead of Proxy at call-time. What do you think?

`AllowAmbiguousTypes` is a bad idea because it _requires_ using `TypeApplications`, but it doesn't tell you _where_ they are required. Using `Proxy` documents exactly where the type annotation is required.

Why are `prove` and `proveWithRepl` polymorphic now?

Reply:
I would argue that `prove @claim ...` is more readable because it's explicit in what happens: we're passing a type to resolve ambiguity. I also suggested we make sure `claim` is the first type argument to make it easy (rather than `prove @_ @claim`).

I find it odd that `prove` does not take any `claim` itself (which would fix all of these problems). Do we have some types which hide `claims` inside existentials or such?

Reply:
it's because we've made the `claimExtractor` part of the `Claim` so now both the extraction and the processing is done in the methods below.

Reply:
Ah so previously we were essentially picking the `claim` type by passing in a concrete `claimExtractor`, but now that we moved that to a type class, the callers of `prove` are no longer picking a `claim` instance so it becomes ambiguous. Is this correct?